feat: More sorting options

pert5432 opened the pull request:
# Ticket(s) Closed

- Closes https://github.com/paradedb/paradedb/issues/1225

## What
Implements returning rows ordered by a field in the index rather than bm25 score

## Why
see issue

## How
Adds `order_by_field` and `order_by_direction` fields to `SearchConfig` and `<index_name>.search` functions. Uses Tantivy `order_by_fast_field` (https://docs.rs/tantivy/latest/tantivy/collector/struct.TopDocs.html#method.order_by_fast_field)

## Tests
Added

pert5432 committed:
order_by params in SearchConfig; ordered searching

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/paradedb/paradedb?pullRequest=1541) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/paradedb/paradedb?pullRequest=1541) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/paradedb/paradedb?pullRequest=1541) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/paradedb/paradedb?pullRequest=1541) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/paradedb/paradedb?pullRequest=1541) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/paradedb/paradedb?pullRequest=1541) it.</sub>

pert5432 commented:
As far as I understand Tantivy doesn't support using `tweak_score` and `order_by_fast_field` at the same time, this means that the sorting isn't stable when using `order_by_field`. 

This means that if somebody makes a search using both `order_by_field` and `stable_sort => true` the sort isn't actually stable, without any indication to the user.

Should I add some validation so these options can't be used together?

rebasedming commented:
> doesn't

Yes, that makes sense to me!

philippemnoel reviewed (commented):
Thanks for making this! Please let us know when it's ready

pert5432 committed:
validate that stable_sort and order_by_field don't get used at the same time

pert5432 committed:
use order_by_u64_field instead

pert5432 committed:
tests

pert5432 committed:
lint: remove static lifetime

pert5432 committed:
Merge branch 'dev' into feat/more-sorting-options

pert5432 commented:
working on resolving the conflicts

pert5432 committed:
rewrite using custom_score

pert5432 committed:
remove unused imports

pert5432 committed:
don't panic when used alongside stable_sort

pert5432 commented:
Rewrote so its compatible with changes from https://github.com/paradedb/paradedb/pull/1556
Since the current implementation sorts in a stable way I removed the `panic!` when using `order_by_field` and `stable_sort` at the same time

philippemnoel commented:
> Rewrote so its compatible with changes from #1556 Since the current implementation sorts in a stable way I removed the `panic!` when using `order_by_field` and `stable_sort` at the same time

Nice! Could you please write documentation for these features? It's in `/docs/`

pert5432 committed:
docs

pert5432 committed:
Merge branch 'dev' into feat/more-sorting-options

pert5432 committed:
improve wording

pert5432 committed:
grammar; clarify stable ordering

philippemnoel commented:
Could you please rebase and we'll review/merge?

eeeebbbbrrrr commented:
Hi @pert5432.  Thanks for this PR.  Since it's been some of my PRs that have been causing the merge conflicts here, I'm going to take responsibility and take over this PR to take it over the finish line.  I think it's basically there, but I have a vision for `scan.rs`.  We'll make sure you get proper credit for your work!  ;)
